### PR TITLE
Add middleware per `Router`

### PR DESCRIPTION
**Pull request by Kludex**

### CODE REVIEW

File: starlette/routing.py [hunk 1]
```diff
@@ -650,6 +652,11 @@ def __init__(
         else:
             self.lifespan_context = lifespan
 
+        self.middleware_stack = self.app
```
Comment (Kludex) [inline starlette/routing.py:673]: I know you'll comment about the names, but I prefer them.

**Review by iudeen — APPROVED** [review]
Looks good.

Do we need to update docs?

**Comment by Kludex** [comment]
@encode/maintainers someone else wants to review this?

File: starlette/routing.py [hunk 2]
```diff
@@ -615,6 +615,8 @@ def __init__(
         # the generic to Lifespan[AppType] is the type of the top level application
         # which the router cannot know statically, so we use typing.Any
         lifespan: typing.Optional[Lifespan[typing.Any]] = None,
+        *,
+        middleware: typing.Optional[typing.Sequence[Middleware]] = None,
```
Comment (abersheeran) [inline starlette/routing.py:627]: Maybe use name `middlewares`?

Comment (Kludex) [reply starlette/routing.py:627]: But we use `middleware` everywhere already...

Comment (abersheeran) [reply starlette/routing.py:627]: 😂 Oh, I didn't even notice that.

**Review by abersheeran — APPROVED** [review]
Simple and elegant modification

**Pull request merged**